shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/CardsProcessingService.java

```diff
@@ -77,84 +80,132 @@ public void applyStudyCard(StudyCard studyCard, List<DatasetAcquisition> acquisi
+                } catch (EntityNotFoundException e) {} // too bad
+            }
+            List<DatasetAcquisition> acquisitions = examination.getDatasetAcquisitions();
+            if (CollectionUtils.isNotEmpty(acquisitions)) {
+                LOG.debug(acquisitions.size() + " acquisitions found for examination with id: " + examination.getId());
+                LOG.debug(qualityCard.getRules().size() + " rules found for study card with id: " + qualityCard.getId() + " and name: " + qualityCard.getName());
+                long rulesStartTs = new Date().getTime();
+                for (QualityExaminationRule rule : qualityCard.getRules()) {
+                    rule.apply(examination, result, downloader);
                 }
-			};
-			//result.removeUnchanged(study);
+                LOG.debug("Quality check for examination " + examination.getId() + " : rules application took " + (new Date().getTime() - rulesStartTs) + "ms");
+            }
 			if (updateTags) {
 			    try {
 			        subjectStudyService.update(result.getUpdatedSubjectStudies());
 			    } catch (EntityNotFoundException e) {
-			        throw new IllegalStateException("Could not update subject-studies", e);
+                    throw new IllegalStateException("Could not update subject-studies", e);
 			    }	    
 			}
+            LOG.info("Quality check for examination " + examination.getId() + " finished in " + (new Date().getTime() - startTs) + " ms");
+            return result;
+		} else {
+			throw new RestClientException("Study card used with emtpy rules.");
+		}
+	}
+
+	/**
+	 * Study cards for quality control: apply on entire study.
+	 * 
+	 * @param studyCard
+	 * @throws MicroServiceCommunicationException 
+	 */
+	public QualityCardResult applyQualityCardOnStudy(QualityCard qualityCard, boolean updateTags, Integer start, Integer stop) throws MicroServiceCommunicationException {
+        long startTs = new Date().getTime();
+        if (qualityCard == null) throw new IllegalArgumentException("qualityCard can't be null");
+        ShanoirEvent event = new ShanoirEvent(ShanoirEventType.CHECK_QUALITY_EVENT, null, KeycloakUtil.getTokenUserId(), "Quality check started on study " + qualityCard.getStudyId() , 4);
+		eventService.publishEvent(event);
+        Study study = studyService.findById(qualityCard.getStudyId());
+		if (study == null ) throw new IllegalArgumentException("study can't be null");
+		if (qualityCard.getStudyId() != study.getId()) throw new IllegalStateException("study and studycard ids don't match");
+		if (CollectionUtils.isNotEmpty(qualityCard.getRules())) {	    
+		    QualityCardResult result = new QualityCardResult();
+            Float i = 0f;
+            List<Examination> examinations;
+            if (start != null && stop != null) {
+                examinations = study.getExaminations().subList(start, stop < study.getExaminations().size() ? stop : study.getExaminations().size());
```

Comment:
Hi @julien-louis, when you select a range of 10 exams, you will always get all from the database. A dedicated call to the database might be much more efficient? With kind regards, Michael

Reply:
so, I mean you call all, and the you select a subList from all, why not only getting the range from the db

Reply:
hm, please check with the debugger and the sql command trace, what is really send to the database, as subList is a collection command, that has nothing to do with hibernate, I would be really surprised, if it ends up in a ranged sql query... but, it would be great if :-)